marshal/cluster.go

```diff
@@ -0,0 +1,272 @@
+		newTopics[topic.Name] = len(topic.Partitions)
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.topics = newTopics
+	return nil
+}
+
+// getOffsetCoordinator returns a kafka.OffsetCoordinator for a specific group.
+func (c *cluster) getOffsetCoordinator(groupID string) (kafka.OffsetCoordinator, error) {
+	return c.broker.OffsetCoordinator(
+		kafka.NewOffsetCoordinatorConf(groupID))
+}
+
+// getClaimPartition calculates which partition a topic should use for coordination. This uses
+// a hashing function (non-cryptographic) to predictably partition the topic space.
+func (c *cluster) getClaimPartition(topicName string) int {
+	// We use MD5 because it's a fast and good hashing algorithm and we don't need cryptographic
+	// properties. We then take the first 8 bytes and treat them as a uint64 and modulo that
+	// across how many partitions we have.
+	hash := md5.Sum([]byte(topicName))
+	uval := binary.LittleEndian.Uint64(hash[0:8])
+	return int(uval % uint64(c.partitions))
+}
+
+// getPartitionState returns a topicState and possibly creates it and the partition state within
+// the State.
+func (c *cluster) getPartitionState(groupID, topicName string, partID int) *topicState {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	group, ok := c.groups[groupID]
+	if !ok {
+		group = make(map[string]*topicState)
+		c.groups[groupID] = group
+	}
+
+	topic, ok := group[topicName]
+	if !ok {
+		topic = &topicState{
+			claimPartition: c.getClaimPartition(topicName),
+			partitions:     make([]PartitionClaim, partID+1),
+		}
+		group[topicName] = topic
+	}
+
+	// Take the topic lock if we can
+	topic.lock.Lock()
+	defer topic.lock.Unlock()
+
+	// They might be referring to a partition we don't know about, maybe extend it
+	if len(topic.partitions) < partID+1 {
+		for i := len(topic.partitions); i <= partID; i++ {
+			topic.partitions = append(topic.partitions, PartitionClaim{})
+		}
+	}
+
+	return topic
+}
+
+// getTopics returns the list of known topics.
+func (c *cluster) getTopics() []string {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	topics := make([]string, 0, len(c.topics))
+	for topic := range c.topics {
+		topics = append(topics, topic)
+	}
+	return topics
+}
+
+// getTopicPartitions returns the count of how many partitions are in a given topic. Returns 0 if a
+// topic is unknown.
+func (c *cluster) getTopicPartitions(topicName string) int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	count, _ := c.topics[topicName]
+	return count
+}
+
+// waitForRsteps is used by the test suite to ask the rationalizer to wait until some number
+// of events have been processed. This also returns the current rsteps when it returns.
+func (c *cluster) waitForRsteps(steps int) int {
+	for {
+		cval := atomic.LoadInt32(c.rsteps)
+		if cval >= int32(steps) {
+			return int(cval)
+		}
+		time.Sleep(5 * time.Millisecond)
+	}
+}
+
+// terminate is called when we're done with the marshaler and want to shut down.
+func (c *cluster) terminate() {
```

Comment:
I think this needs to remove this cluster from the global clusters map. Otherwise if we encounter a problem and get terminated, we can never start another cluster pointed at that set of brokers.